larq_compute_engine/core/ruy_pack.h

```diff
@@ -0,0 +1,137 @@
+ * that if we rely on the Ruy packing code we end up falling back to the slow
+ * reference implementation that copies each element one-by-one.
+ *
+ * In our optimised LCE kernels, the LHS matrix is the weights and the RHS is
+ * the input activations. As the packed weights can be computed once then
+ * cached, we only care about optimising for packing the RHS. In all our
+ * optimised kernels, the RHS has a 4x4 column-major kernel layout; we therefore
+ * extend the Ruy packing with an optimised (but portable, C++) packing routine
+ * for 4x4 column-major kernel layouts.
+ */
+
+#include "larq_compute_engine/core/types.h"
+#include "ruy/pack_common.h"
+
+namespace compute_engine {
+namespace tflite {
+
+using compute_engine::core::TBitpacked;
+using namespace ruy;
+
+template <Path ThePath, typename FixedKernelLayout>
+struct LceRuyPackImpl
+    : PackImpl<ThePath, FixedKernelLayout, TBitpacked, TBitpacked, TBitpacked> {
+};
+
+template <Path ThePath>
+struct LceRuyPackImpl<ThePath, FixedKernelLayout<Order::kColMajor, 4, 4>> {
+  // This method is derived Ruy's 16x4 column-major int8 packing routine:
+  // https://github.com/google/ruy/blob/c9f5f9cecde3d6314df6e7d91517356bf07135eb/ruy/pack_arm.h#L122-L197
+  static void Run(Tuning, const Mat<TBitpacked>& src_matrix,
+                  PMat<TBitpacked>* packed_matrix, int start_col, int end_col) {
+    profiler::ScopeLabel label("Pack (ColMajor, 4x4)");
+
+    // Ruy supports collecting column sums during the packing process, which we
+    // have no need for.
+    RUY_DCHECK_EQ(packed_matrix->sums, nullptr);
+
+    // Likewise, Ruy supports arbitrary zero points, but we only use true-zero.
+    RUY_DCHECK_EQ(src_matrix.zero_point, 0);
+
+    RUY_DCHECK(IsColMajor(src_matrix.layout));
+    RUY_DCHECK(IsColMajor(packed_matrix->layout));
+    RUY_DCHECK_EQ(start_col % 4, 0);
+
+    TBitpacked zerobuf[4];
+    memset(zerobuf, 0, sizeof(zerobuf));
+
+    for (int block_col = start_col; block_col < end_col; block_col += 4) {
+      int src_stride = src_matrix.layout.stride;
+      const TBitpacked* src_ptr0 =
+          src_matrix.data.get() + src_stride * block_col;
+      const TBitpacked* src_ptr1 = src_ptr0 + src_stride;
+      const TBitpacked* src_ptr2 = src_ptr1 + src_stride;
+      const TBitpacked* src_ptr3 = src_ptr2 + src_stride;
+      int src_inc0 = 4;
+      int src_inc1 = 4;
+      int src_inc2 = 4;
+      int src_inc3 = 4;
+      if (block_col >= src_matrix.layout.cols - 3) {
+        if (block_col >= src_matrix.layout.cols - 0) {
+          src_ptr0 = zerobuf;
+          src_inc0 = 0;
+        }
+        if (block_col >= src_matrix.layout.cols - 1) {
+          src_ptr1 = zerobuf;
+          src_inc1 = 0;
+        }
+        if (block_col >= src_matrix.layout.cols - 2) {
+          src_ptr2 = zerobuf;
+          src_inc2 = 0;
+        }
+        if (block_col >= src_matrix.layout.cols - 3) {
+          src_ptr3 = zerobuf;
+          src_inc3 = 0;
+        }
```

Comment:
If I understand the code correctly, this condition should always be true.
```suggestion
        src_ptr3 = zerobuf;
        src_inc3 = 0;
```